### PR TITLE
feat: extract and surface AI model and version

### LINKED ISSUE #29: Extract and surface AI tool type, Model used, and version

### PR DESCRIPTION
**Pull request by abhinavgautam01**
fixes #29

### Description
Previously, all detection data was lumped into a single `Detail` string, making it difficult to systematically identify the exact AI models and versions being used. This PR introduces dedicated fields for `Model` and `Version` in our core detection structures and surfaces them natively in the CLI outputs.

### Changes Made
- **Core Update**: Added `Model` and `Version` fields to the `detection.Finding` struct.
- **GitNotes Extractor**: Updated `detection/gitnotes/gitnotes.go` to directly capture the model from `prompt.AgentID.Model` instead of burying it in the detail string.
- **Co-Author Extractor**: Added regex parsing in `detection/coauthor/coauthor.go` to pull out explicit model strings from `Co-Authored-By` trailers (e.g. `aider (gpt-4o)` now captures `gpt-4o`, `Claude Opus 4` captures `Claude Opus 4`, etc.).
- **Output Rendering**: Updated the CLI's standard text formatter (`output/output.go`) to automatically append `[model_name]` next to the tool name if a model was successfully extracted.
- **Testing**: Updated `coauthor_test.go` to rigorously assert model extractions and verified all project tests pass natively.

### Testing
- `go test ./...` passes without errors.
- Text outputs now neatly group tools alongside their respective models (e.g. `Aider [gpt-4o]`).

### CODE REVIEW

File: detection/coauthor/coauthor_test.go [hunk 1]
```diff
@@ -9,78 +9,93 @@ import (
...
+			wantModels: []string{"Claude Sonnet 4"},
 		},
 		{
-			name:      "Cursor trailer",
-			message:   "refactor: extract method\n\nCo-Authored-By: Cursor <cursoragent@cursor.com>",
-			wantTools: []string{"Cursor"},
+			name:       "Cursor trailer",
+			message:    "refactor: extract method\n\nCo-Authored-By: Cursor <cursoragent@cursor.com>",
+			wantTools:  []string{"Cursor"},
+			wantModels: []string{"Cursor"},
```
Comment (MoralCode) [inline detection/coauthor/coauthor_test.go]: i feel like i would expect the model to be empty here since i would consider cursor a tool through which models like composer 2.5 can be run.

File: detection/coauthor/coauthor_test.go [hunk 2]
```diff
@@ -9,78 +9,93 @@ import (
...
+			wantModels: []string{"gpt-4o"},
 		},
 		{
-			name:      "Aider trailer with different model",
-			message:   "feat: add endpoint\n\nCo-Authored-By: aider (claude-3.5-sonnet) <noreply@aider.chat>",
-			wantTools: []string{"Aider"},
+			name:       "Aider trailer with different model",
+			message:    "feat: add endpoint\n\nCo-Authored-By: aider (claude-3.5-sonnet) <noreply@aider.chat>",
+			wantTools:  []string{"Aider"},
+			wantModels: []string{"claude-3.5-sonnet"},
```
Comment (MoralCode) [inline detection/coauthor/coauthor_test.go]: should we try to standardize the formatting of these model names so theyre easier for consumers to parse? maybe separating out the version (3.5 sonnet in this case) into a new field this would be better for a followup PR?

**Review by MoralCode — COMMENTED** [review]
I was skeptical at first given this is a new PR filed so close time-wise to your other one, but overall I like how small and concise the diff is.

Leaving this as comments to see if other contributors/maintainers have feedback, but overall i think this is a good change. I'd like to see maybe more robust parsing that isnt just focused on two frontier models, but id probably be fine with iterating on that over time

File: detection/coauthor/coauthor_test.go (hunk 1 above)
Comment (abhinavgautam01) [reply detection/coauthor/coauthor_test.go]: ohh, my bad...i havve just pushed a fix to this PR that ensures the Model  field remains cleanly empty when Cursor is detected...

File: detection/coauthor/coauthor_test.go (hunk 2 above)
Comment (abhinavgautam01) [reply detection/coauthor/coauthor_test.go]: this is a fantastic idea, but since AI model naming conventions are notoriously messy across vendors (e.g.  gpt-4o  vs  claude-3.5-sonnet  vs  Claude Opus 4 ), building a robust regex or registry to cleanly separate the "Model Name" from the "Version" everywhere is going to be complex...
i definitely agree with you that this is better suited for a dedicated follow-up PR where we can properly design and test the version extraction logic...

**Review by andrew — CHANGES_REQUESTED** [review]
Please rebase and port the co-author changes into the current `detection/trailer` package. The old `detection/coauthor` package and its tests now have modify/delete conflicts with `main`.

Please also preserve every distinct tool/model pair. Both the co-author and git-notes paths currently deduplicate only by tool, so later models used through the same tool are discarded. For git notes, map iteration also makes the surviving model unpredictable. Add multi-model tests for both sources.

Finally, either populate `Finding.Version` or leave the field and renderer out until version extraction is implemented. No detector currently assigns it.

**Comment by abhinavgautam01** [comment]
Thanks, addressed these.

- Rebasing onto current main resolved the old `detection/coauthor` modify/delete conflicts and ports the co-author model extraction into `detection/trailer`.
- Co-author and git-notes detection now deduplicate by distinct tool/model pairs, so multiple models for the same tool are preserved.
- Git-notes prompt handling is deterministic by sorting prompt IDs before emitting findings.
- Added multi-model tests for both trailer/coauthor and git-notes detection.
- Removed the unimplemented `Finding.Version` field/rendering for now.

File: detection/trailer/trailer.go [hunk 3]
```diff
@@ -94,6 +94,31 @@ type Detector struct{}
...
+type toolModelPair struct {
+	tool  string
+	model string
+}
+
+func extractCoauthorModel(tool, namePart string) string {
+	namePart = strings.TrimSpace(namePart)
+
+	switch tool {
+	case "Aider":
```
Comment (omkar-foss) [inline detection/trailer/trailer.go]: It'll be great if this function to extract model from the round brackets could support all tool types supported by this detector, and not just Aider and Claude Code. I suppose that'd provide good amount of consistency to this feature.

File: detection/trailer/trailer_test.go [hunk 4]
```diff
@@ -12,61 +12,85 @@ func TestDetect(t *testing.T) {
...
 		wantTools      []string
+		wantModels     []string
 		wantConfidence []detection.Confidence
 	}{
 		// Co-Authored-By tests start here
 		{
 			name:           "coauthor: Claude trailer with Opus model",
 			message:        "fix: update handler\n\nCo-Authored-By: Claude Opus 4 <noreply@anthropic.com>",
 			wantTools:      []string{"Claude Code"},
+			wantModels:     []string{"Claude Opus 4"},
```
Comment (omkar-foss) [inline detection/trailer/trailer_test.go]: This can simply be Opus 4 as model and version are being surfaced here and we already have the corresponding tool name above i.e. Claude Code.

File: output/output.go [hunk 5]
```diff
@@ -72,6 +72,13 @@ func FormatJSONFindings(w io.Writer, findings []detection.Finding) error {
 	}{Findings: findings})
 }
 
+func formatTool(f detection.Finding) string {
```
Comment (omkar-foss) [inline output/output.go]: This function can be a method belonging to detection.Finding, and it should handle f.Tool being empty when f.Model is non-empty, as an edge case.

**Review by omkar-foss — CHANGES_REQUESTED** [review]
PR looks fine overall, added a few comments. Also would be good to have more test cases.

File: detection/trailer/trailer.go (hunk 3 above)
Comment (abhinavgautam01) [reply detection/trailer/trailer.go]: I updated this so parenthesized model extraction is handled generically before tool-specific fallbacks. Cursor, Copilot, Aider, Claude Code, etc. now support values like `Tool (model)` consistently.

File: detection/trailer/trailer_test.go (hunk 4 above)
Comment (abhinavgautam01) [reply detection/trailer/trailer_test.go]: I updated Claude Code co-author extraction to strip the duplicated `Claude` prefix, so `Co-Authored-By: Claude Opus 4 ...` now surfaces tool `Claude Code` and model `Opus 4`.

File: output/output.go (hunk 5 above)
Comment (abhinavgautam01) [reply output/output.go]: I moved this into `detection.Finding.DisplayTool()` and added handling for model-only findings, so text output now uses the shared method instead of an output-local helper.

**Comment by abhinavgautam01** [comment]
@omkar-foss  thanks for the review. I addressed the comments and added tests around generic parenthesized model extraction, Claude model normalization and model-only display formatting.

**Review by andrew — APPROVED** [review]
The requested changes are addressed.

**Comment by omkar-foss** [comment]
> @omkar-foss thanks for the review. I addressed the comments and added tests around generic parenthesized model extraction, Claude model normalization and model-only display formatting.

Thanks for making the changes @abhinavgautam01, looks good to me. I was going to approve and merge but looks like @andrew beat me to it, he's on a roll :P
